### PR TITLE
[MPS] Dispatch outer bin edges selection function

### PR DESCRIPTION
**Pull request by qqaatw**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #101792
* #101692

Dispatch the selection function to prevent using `is_mps()` in `Histogram.cpp`.

### <samp>🤖 Generated by Copilot at b329a02</samp>

This pull request refactors and implements the logic for inferring the bin edges of histograms from the input tensor for different device types. It introduces a dispatch stub `histogram_select_outer_bin_edges_stub` and moves the device-specific code to separate files, such as `HistogramKernel.cpp` and `HistogramKernel.mm`. This improves the modularity and readability of the histogram functions.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Comment by qqaatw** [comment]
@kulinseth @malfet Could you review this stack?

**Review by malfet — CHANGES_REQUESTED** [review]
Please add test (i.e. if this op is not yet covered by OpInfo, please feel free to add OpInfo test in separate PR on stack)

File: aten/src/ATen/native/mps/operations/HistogramKernel.mm [hunk 1]
```diff
@@ -396,6 +397,30 @@ static void histogramdd_linear_kernel(const Tensor& self,
...
+    leftmost_edges[i] = min[i].item().to<scalar_t>();
+    rightmost_edges[i] = max[i].item().to<scalar_t>();
+  }
+}
+
+static void histogram_select_outer_bin_edges_kernel(const Tensor& input,
+                                                    const int64_t N,
+                                                    std::vector<double>& leftmost_edges,
+                                                    std::vector<double>& rightmost_edges) {
+  AT_DISPATCH_FLOATING_TYPES(input.scalar_type(), "histogramdd", [&]() {
```
[the remark this answers is not in the document]
Comment (qqaatw) [reply aten/src/ATen/native/mps/operations/HistogramKernel.mm:418]: It wouldn't as the dispatched scalar_t is only used for casting the resulting c++ type from .item(). Though, I think we can remove the dtype dispatch here at all (looks unnecessary to mps), I will update it.

**Comment by qqaatw** [comment]
Gently pinging @malfet

**Comment by qqaatw** [comment]
@malfet can you please take a look? the failing tests are from broken trunk I think.

**Comment by qqaatw** [comment]
Hey @malfet, can you please review?

**Comment by qqaatw** [comment]
Ping @malfet for review, thanks!

**Review by albanD — APPROVED** [review]
Thanks for the cleanup!

**Pull request closed without merging**